File: lib/spring-ts/artifacts/phase3-agent-a8/measure-baseline.mjs
#!/usr/bin/env node
/**
 * artifacts/phase3-agent-a8/measure-baseline.mjs
 *
 * Source-side metric capture for Phase 3 Agent A8 (expression_children polish).
 * Counts repetition phrases, brief headline >28자 violations, and other anchors
 * across all expression_children fragments. Output: stable JSON suitable for
 * before/after diff in artifacts/phase3-agent-a8/.
 */
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const SPRING_TS_ROOT = path.resolve(__dirname, '../..');
const NARRATIVE_DIR = path.join(SPRING_TS_ROOT, 'data/narrative/expression_children');

const TRACKED_PHRASES = [
  '결이',
  '결을',
  '결이에요',
  '흐름',
  '한 박자',
  '페이스',
  '자기 색',
  '다음 세대',
  '단정',
  '단정 없이',
  '기록',
  '천천히',
  '풀려',
  '결실',
  '결로',
  '결과',
  '매듭',
  '편이',
  '또렷',
  '단단',
];

function walk(dir, files = []) {
  for (const f of fs.readdirSync(dir)) {
    const fp = path.join(dir, f);
    const st = fs.statSync(fp);
    if (st.isDirectory()) walk(fp, files);
    else if (f.endsWith('.fragments.json')) files.push(fp);
  }
  return files;
}

function tokensToText(tokens) {
  return (tokens || [])
    .map((t) => {
      if (t.kind === 'text') return t.value;
      if (t.kind === 'tag') return t.label || '';
      return '';
    })
    .join('');
}

function countKoreanChars(text) {
  return Array.from(text).filter((c) => /[가-힣]/.test(c)).length;
}

function main() {
  const files = walk(NARRATIVE_DIR);
  const summary = {
    totalFragments: 0,
    totalBriefFragments: 0,
    briefHeadlineViolations: 0,
    violationDetails: [],
    repetitionCounts: {},
    gatingCoverage: {},
    expertTagSet: new Set(),
  };
  for (const p of TRACKED_PHRASES) summary.repetitionCounts[p] = 0;

  for (const file of files) {
    const data = JSON.parse(fs.readFileSync(file, 'utf-8'));
    for (const frag of data.fragments) {
      summary.totalFragments += 1;
      const text = tokensToText(frag.templateTokens);
      for (const p of TRACKED_PHRASES) {
        const re = new RegExp(p, 'g');
        const matches = text.match(re);
        if (matches) summary.repetitionCounts[p] += matches.length;
      }
      if (frag.axis?.depth === 'brief') {
        summary.totalBriefFragments += 1;
        const len = countKoreanChars(text);
        if (len > 28) {
          summary.briefHeadlineViolations += 1;
          summary.violationDetails.push({ id: frag.fragmentId, len, text });
        }
      }
      if (frag.axis?.depth === 'expert') {
        for (const t of frag.tags || []) summary.expertTagSet.add(t);
      }
      const gKey = Object.keys(frag.gating || {}).sort().join('+') || 'wildcard';
      summary.gatingCoverage[gKey] = (summary.gatingCoverage[gKey] || 0) + 1;
    }
  }
  summary.expertTagSet = Array.from(summary.expertTagSet).sort();
  const out = {
    fragmentFiles: files.length,
    totalFragments: summary.totalFragments,
    totalBriefFragments: summary.totalBriefFragments,
    briefHeadlineViolations: summary.briefHeadlineViolations,
    violationDetails: summary.violationDetails,
    repetitionCounts: summary.repetitionCounts,
    gatingCoverage: summary.gatingCoverage,
    expertTagDiversity: summary.expertTagSet.length,
    expertTagSet: summary.expertTagSet,
  };
  const outputArg = process.argv[2];
  if (outputArg) {
    const outputPath = path.isAbsolute(outputArg) ? outputArg : path.join(__dirname, outputArg);
    fs.writeFileSync(outputPath, JSON.stringify(out, null, 2) + '\n');
    console.error(`Wrote summary to ${outputPath}`);
  }
  console.log(JSON.stringify(out, null, 2));
}

main();

File: lib/spring-ts/artifacts/phase3-agent-a6/measure_baseline.py
"""Phase 3 A6 baseline measurement for career narratives."""
import json
import os
import sys

ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
career_dir = os.path.join(ROOT, 'data', 'narrative', 'career')

results = {
    'fragments': 0,
    'brief_violations': [],
    'word_counts': {'결이': 0, '결을': 0, '결에': 0, '결로': 0, '결은': 0, '결의': 0, '흐름': 0, '한 박자': 0, '페이스': 0, '자리': 0},
    'gating_axes': {},
    'tags_used': {},
    'by_period': {},
    'by_depth': {},
}

for root, dirs, files in os.walk(career_dir):
    for fn in files:
        if fn.endswith('.fragments.json'):
            path = os.path.join(root, fn)
            with open(path, 'r', encoding='utf-8') as f:
                bundle = json.load(f)
            for frag in bundle['fragments']:
                results['fragments'] += 1
                text = ''
                for tok in frag.get('templateTokens', []):
                    if tok['kind'] == 'text':
                        text += tok.get('value', '')
                    elif tok['kind'] == 'tag':
                        text += tok.get('label', '')
                for word in results['word_counts']:
                    results['word_counts'][word] += text.count(word)
                period = frag['axis']['period']
                depth = frag['axis']['depth']
                results['by_period'][period] = results['by_period'].get(period, 0) + 1
                results['by_depth'][depth] = results['by_depth'].get(depth, 0) + 1
                if depth == 'brief':
                    hangul_count = sum(1 for c in text if '가' <= c <= '힣')
                    if hangul_count > 28:
                        results['brief_violations'].append({
                            'id': frag['fragmentId'], 'len': hangul_count, 'text': text
                        })
                gating = frag.get('gating', {})
                gkey = '|'.join(sorted(gating.keys())) if gating else 'wildcard'
                results['gating_axes'][gkey] = results['gating_axes'].get(gkey, 0) + 1
                for t in frag.get('tags', []):
                    results['tags_used'][t] = results['tags_used'].get(t, 0) + 1

total_kyeol = sum(results['word_counts'][k] for k in results['word_counts'] if k.startswith('결'))
print(f'Total fragments: {results["fragments"]}')
print(f'By period: {results["by_period"]}')
print(f'By depth: {results["by_depth"]}')
print(f'\nBrief violations (>28 chars): {len(results["brief_violations"])}')
for v in results['brief_violations']:
    print(f'  {v["id"]} len={v["len"]}: {v["text"]}')
print(f'\nWord counts:')
for w, c in results['word_counts'].items():
    print(f'  {w}: {c}')
print(f'Total "결*" forms: {total_kyeol}')
print(f'\nGating axes ({len(results["gating_axes"])} unique):')
for k, v in sorted(results['gating_axes'].items(), key=lambda x: -x[1]):
    print(f'  {k}: {v}')
print(f'\nTags used ({len(results["tags_used"])}):')
for t, c in sorted(results['tags_used'].items(), key=lambda x: -x[1]):
    print(f'  {t}: {c}')

if '--write' in sys.argv:
    out = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'baseline.json')
    with open(out, 'w', encoding='utf-8') as f:
        json.dump(results, f, ensure_ascii=False, indent=2)
    print(f'\nBaseline written to {out}')
